### PR TITLE
feat: Add ticket chart analytics

### LINKED ISSUE #261: Show graph indicating ticket sales
Show graph indicating ticket sales on daily basis

Taking this up

### PR DESCRIPTION
**Pull request by iamareebjamal**
Fixes #261

- Add chart view for ticket analytics
- Add chart mutator for handling analytics logic

device-2017-07-09-000009

### CODE REVIEW

**Comment by iamareebjamal** [comment]
This is a mock representation. Need to discuss the analysis implementation @mejariamol

**Comment by mejariamol** [comment]
@iamareebjamal first will have to decide what data will be shown in the chart. that is what will be on y axis.

**Comment by iamareebjamal** [comment]
This is per day ticket sale analytics. Y axis is amount of tickets bought and X axis is the date when ticket was bought

**Comment by mejariamol** [comment]
I think we should have a separate chart for paid tickets with total earn on the Y axis. What do you say? As for paid tickets amount will be more informative that just a number. @iamareebjamal

**Comment by iamareebjamal** [comment]
That will be implemented in Tickets Page #267

**Comment by mejariamol** [comment]
Okay. So for implementation of sales versus time, we will need a sorted Ticket list with time of purchase

It can be achieved by sorting Attendee list with `Order` using `completedAt` attribute.  and creating sorted ticket list by iterating through these sorted attendees.

**Comment by iamareebjamal** [comment]
That was what I was thinking. Instead of iterating through tickets, we should iterate through orders. Because orders are less, and are done on a single date. What do you think

**Comment by mejariamol** [comment]
We will need:
1. ticket type( from Ticket)
2. completed at (from order)
3. quantity (from order)

**Comment by mejariamol** [comment]
Just have to think how can we get this info sorted with `completedAt`

**Comment by iamareebjamal** [comment]
We don't need to sort at all, this will be handled by the charting library

**Comment by mejariamol** [comment]
Ok, so what values does charting library take?

**Comment by iamareebjamal** [comment]
It takes a point (X, Y) which in our case will be (purchased, date)

**Comment by mejariamol** [comment]
but order and ticket are related by an attendee. How can we get ticket type by iterating through orders?

**Comment by iamareebjamal** [comment]
order will have one to many relationship with attendees

**Comment by mejariamol** [comment]
Yeah. then it is better to iterate through orders.

**Comment by mejariamol** [comment]
Once we should check complexities for both approaches

**Comment by mejariamol** [comment]
Orders = m
attendees = n
iterating through to orders will have m X n
but through attendees it will have just n

Is is correct? :smile:

**Comment by iamareebjamal** [comment]
No, because we have to get the order from attendees as well, and orders will be repeated, so 20 attendees will have the same order

And if we parse by order, then total attendees will be parsed only for their count and type, but it'll be done in 1 pass only

File: app/src/main/java/org/fossasia/openevent/app/event/detail/ChartMutator.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    private LineSet paidSet = new LineSet();
+    private LineSet donationSet = new LineSet();
+
+    protected ChartMutator(IUtilModel utilModel) {
+        this.utilModel = utilModel;
+
+        gridPaint.setColor(Color.parseColor("#992ecc71"));
+        gridPaint.setStrokeWidth(5);
+
+        freeSet = new LineSet(new String[] {"Jan", "Feb", "Mar"}, new float[] {1.2f, 3.6f, 4.5f});
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/event/detail/ChartMutator.java]: Seems like we will have to calculate month wise tickets separately. ?

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/event/detail/ChartMutator.java]: This is just mock data, we won't do month wise analysis, we will do aggregate analysis of total sales over the time

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/event/detail/ChartMutator.java]: Ok. So what will be our time interval?

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/event/detail/ChartMutator.java]: No time interval, just date by sales mapping, it'll be automatically handles by the graphing library

**Comment by iamareebjamal** [comment]
device-2017-07-19-095615

@mejariamol Please review

**Comment by iamareebjamal** [comment]
@mejariamol Can you please check now. I accidentally used Java 8 forEach construct

**Comment by mejariamol** [comment]
@iamareebjamal Some suggestions:
- please use same colors in the chart and for progress bars.
- Currently, it is not clear what actually the chart shows. Please add required info at axes.
- please take care of extremity:
  - what will be shown when no tickets are sold

**Comment by iamareebjamal** [comment]
Same colors are used in both cases. 

Axes will not give an immersive experience, and tooltips will handle the data information. The chart is a standard sales/time which is quite known form of analytic chart.

About extremities, I don't think the app will crash, just show an empty chart. I will check it

**Comment by mejariamol** [comment]
>Axes will not give an immersive experience, and tooltips will handle the data information. The chart is a standard sales/time which is quite known form of analytic chart.

but at least it should show values at axes. :smile:

**Comment by iamareebjamal** [comment]
@mejariamol Imagine an event with 1000 tickets. How will the axes look then? We don't have logic to dynamically show the appropriate amount of labels on the axis. Probably in future

About the colors, I have used primary colors for lines and for dots, I have used the light color. The shade is used in progress bars to show complete incomplete data. What else should be done here?

**Comment by mejariamol** [comment]
>We don't have logic to dynamically show the appropriate amount of labels on the axis.

For now, let's keep the number of labels at axes constant. Their values will be assigned dynamically. For example, 4 labels on the Y axis and 3 labels on the X axis. 
Please try if we can set logic to assign these 7 variables dynamically. Because showing chart without any value at axes is really pointless.

**Comment by mejariamol** [comment]
@iamareebjamal is it possible now to switch chart library to [PhilJay/MPAndroidChart](https://github.com/PhilJay/MPAndroidChart). It is well documented. what do you say? I shall help in switching.

**Comment by iamareebjamal** [comment]
@mejariamol I don't know what such features it brings which WilliamCharts doesn't. I recommend you to check out its sample app https://play.google.com/store/apps/details?id=com.db.williamchartdemo&hl=en

**Comment by iamareebjamal** [comment]
And about the axes. I am not saying that we shouldn't add it. It was never my implication. The chart is far from complete, but if we set out to make it perfect in this PR only, it'll take forever. Issues should be such that they are solved in a day. This already took a lot of time. There are several things remaining on it and should be solved as other issues, so that we can independently and iteratively develop it.

**Comment by mejariamol** [comment]
>I don't know what such features it brings which WilliamCharts doesn't. I recommend you to check out its sample app https://play.google.com/store/apps/details?id=com.db.williamchartdemo&hl=en

yeah. right. WilliamCharts also provides good doc. :+1:

**Comment by mejariamol** [comment]
@iamareebjamal what do you say about [PhilJay/MPAndroidChart](https://github.com/PhilJay/MPAndroidChart)? shall I send PR to your repo? Or will you implement as most of the functions are similar? So it will be easier for you.

**Comment by iamareebjamal** [comment]
I'll do it :+1:

**Comment by iamareebjamal** [comment]
device-2017-07-21-034359

@mejariamol Please review now

**Comment by mariobehling** [comment]
Please resolve conflicts.

**Pull request merged**